### PR TITLE
Add auto-rendered product layout

### LINKED ISSUE #229: Add auto-rendered product layout
Add auto-rendered product layout where the product data is rendered based on [product drop](https://www.voog.com/developers/markup/ecommerce-objects/product) data.

It should look like current "Product" layout. Use new [image](https://www.voog.com/developers/markup/tags/image) or [image_data](https://www.voog.com/developers/markup/tags/image-data) tag for product cover image.

It content type must be `product`:

```json
{
  "content_type": "product",
  "component": false,
  "file": "layouts/auto_generated_product.tpl",
  "layout_name": "auto_generated_product",
  "title": "Auto-generated product"
}
```

See also buy button rendering example in [Nuuk design](https://github.com/Voog/design-nuuk/blob/master/components/buy-button.tpl).

## TODO

* [ ] Add "Auto-generated product" layout
* [ ] Check and improve "[product list](https://www.voog.com/support/online-store/product-list-content-area)" widget styles when needed.

### PR DESCRIPTION
**Pull request by PeeterPaal**
Closes #229 

Add auto-rendered product system layout where the product page is rendered based on product.

Improve scaled images rendering in Chrome (`image-rendering: -webkit-optimize-contrast`).

### CODE REVIEW

File: layouts/auto_generated_product.tpl [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+{% include "product-variables" %}
+
+<html class="layout-root {{ language_flags_mode }} {{ language_names_mode }} {{ language_menu_mode }} {{ site_search_mode }}" lang="{{ page.language_code }}" data-view-state="{{ view_mode }}" data-lingual-mode="{{ site_lingual_mode }}" data-menu-main-state="closed">
+  <head prefix="og: http://ogp.me/ns#">
+    {% include "template-head" layout_product: true %}
+  </head>
+
+  <body class="layout-container item-page">
+
+    {%- capture _button_attributes %}
```
Comment (tanelj) [inline layouts/auto_generated_product.tpl:12]: Its not used here. Move this capture to `buy-button` component.

File: layouts/auto_generated_product.tpl [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    {% include "template-svg-spritesheet" %}
+
+    <div class="layout-wrap">
+      {% include "layout-header" %}
+
+      <main class="layout-body layout-body--narrow">
+
+        <div class="content-inner product-content">
+          <div class="layout-body__content">
+            <div class=" items-body" {{ edy_intro_edit_text }}>
```
Comment (tanelj) [inline layouts/auto_generated_product.tpl:27]: `edy_intro_edit_text` is deprecated and not used any more. Please remove.
```suggestion
            <div class=" items-body">
```

File: layouts/auto_generated_product.tpl [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                {% else %}
+                  {% assign item_image_state = "without-image" %}
+                {% endif %}
+
+                <div class="content-item-box {{ item_image_state }} js-content-item-box">
+                  <div class="item-top">
+                    {%- if productImage != blank -%}
+                      <div class="loader js-loader"></div>
+                      <div class="top-inner aspect-ratio-inner">
+                      {%- assign imageClass = "item-image " | append: "not-cropped " | append: "top-inner" -%}
```
Comment (tanelj) [inline layouts/auto_generated_product.tpl:42]: Since not controlled by user at the moment use simple assign:
```suggestion
                      {%- assign imageClass = "item-image  not-cropped  top-inner" -%}
```

**Review by tanelj — CHANGES_REQUESTED** [review]
Nice, but some feedback!

* [ ] Product image area with image should be same color as the page bg. At the momet:
Screenshot 2022-01-31 at 16 03 51
* [ ] Buy button disable state must be implemente also on edit mode on auto gen page. (Currently is red and clickable).
* [ ] Product without image (without-image) has no placeholder. Use frame and https://github.com/Voog/design-nuuk/blob/master/assets/ico-placeholder.svg

File: sources/stylesheets/content-area/_text.scss [hunk 4]
```diff
@@ -331,6 +331,13 @@
         cursor: default;
         background-color: $color-disabled;
       }
+
+      &.edy-buy-button {
```
Comment (robertpih) [inline sources/stylesheets/content-area/_text.scss:335]: Sisuala abil lisatud ostunuppudel pole hetkel editmodes disabled stiile küljes. Äkki teeks siin ka nii, et ka autogenereeritud lehel on nad editmodes  ikkagi 'orginaalvärvi'. Muidu klient disainitööriistas nupu värve muutes muudatust enne ei näe, kui läheb eelvaatesse.

**Review by robertpih — CHANGES_REQUESTED** [review]
Vinks-vonks, mõned asjad jäid silma.

File: layouts/auto_generated_product.tpl [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    <div class="layout-wrap">
+      {% include "layout-header" %}
+
+      <main class="layout-body layout-body--narrow">
+
+        <div class="content-inner product-content">
+          <div class="layout-body__content">
+            <div class=" items-body">
+              <div class="content-illustrations">
+                {% assign product_image = product.image %}
```
Comment (tanelj) [inline layouts/auto_generated_product.tpl:36]: No need for separate `product_image` variable. Juts use `product.image` directly.

File: layouts/auto_generated_product.tpl [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                  {% assign item_image_state = "without-image" %}
+                {% endif %}
+
+                <div class="content-item-box {{ item_image_state }} js-content-item-box">
+                  <div class="item-top">
+                    {%- if product_image != blank -%}
+                      <div class="loader js-loader"></div>
+                      <div class="top-inner aspect-ratio-inner product-page">
+                        {%- assign image_class = "item-image not-cropped" -%}
+                        {% image product_image target_width: "1280" class: image_class loading: "lazy" %}
```
Comment (tanelj) [inline layouts/auto_generated_product.tpl:50]: Here seems that here is something mixed – JS Lazy Load and IMG `loading="lazy"` attribute.

`<div class="loader js-loader"></div>` should be removed or

image tag must be replaced with `image_data` to use 

```liquid
{%- assign image_class = "item-image js-lazyload not-cropped" -%}
{% image_data productImage target_width="1280" class: imageClass %}
```

**Review by tanelj — CHANGES_REQUESTED** [review]
Nice, but some minor comments.

File: layouts/auto_generated_product.tpl [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+      {% endif %}
+    </div>
+
+    {% include "site-signout" %}
+    {% include "template-javascripts" %}
+    {% include "template-tools" %}
+    {% include "content-item-tools" %}
+
+    <script>
+      template.handleProductPageContent();
```
Comment (pxska) [inline layouts/auto_generated_product.tpl:115]: Siin peaks kontrollima ka seda, et kas `template` muutuja on olemas.

**Review by pxska — CHANGES_REQUESTED** [review]

File: layouts/auto_generated_product.tpl [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    {% endunless %}
+  {% endcapture %}
+
+  {% capture bottom_content_size %}
+    {{ bottom_content_html | size | minus: 1 }}
+  {% endcapture %}
+
+  {% unless bottom_content_size contains "-" %}
+    {% assign bottom_content_has_content = true %}
+  {% endunless %}
```
Comment (pxska) [inline layouts/auto_generated_product.tpl:22]: This is always `true`, because the `capture` block also takes into account the whitespace the tags generate, hence the size will not be 0.

This can also be simplified by writing it like this:
```liquid
{% capture bottom_content_html %}
  {% unless editmode %}
    {% content bind=product name="content" %}
  {% endunless %}
{% endcapture %}

{% assign bottom_content_size = bottom_content_html | strip | size %}
```
and whenever you need to check if the bottom_content has content, you just do `if bottom_content_size > 0`.

File: layouts/auto_generated_product.tpl [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                    {%- if product.image != blank -%}
+                      <div class="top-inner aspect-ratio-inner product-page">
+                        {%- assign image_class = "item-image not-cropped" -%}
+                        {% image product.image target_width: "1280" class: image_class loading: "lazy" %}
+                      </div>
+                    {%- endif -%}
+                  </div>
+                </div>
+                <div class="content-gallery content-area js-product-gallery" data-search-indexing-allowed="true">
+                  {% content bind=product name="gallery" %}
```
Comment (pxska) [inline layouts/auto_generated_product.tpl:54]: We should also check if this content area has any content (just like the bottom content), because if this is empty, an empty `div` element will be rendered:
```liquid
{% capture gallery_content_html %}
  {% unless editmode %}
    {% content bind=product name="gallery" %}
  {% endunless %}
{% endcapture %}

{% assign gallery_content_size = gallery_content_html | strip | size %}
```

File: sources/javascripts/concat/base.js [hunk 10]
```diff
@@ -537,6 +537,38 @@
...
+    }, 25));
+
+    var changeProductImagePos = function () {
+      var productGallery = $('.js-product-gallery');
+      var productImageContentBox = $('.js-content-item-box');
+      var buyBtnContent = $('.js-buy-btn-content');
+
+      if ($('.js-buy-btn-content .edy-buy-button-container').length >= 1) {
+        if ($(window).width() <= 640) {
+          if ($('.js-buy-btn-content + .js-product-gallery').length <= 0) {
```
Comment (pxska) [inline sources/javascripts/concat/base.js:560]: It seems that the `length` cant be under 0 in any way, so the check should be like this:
```js
if ($('.js-buy-btn-content + .js-product-gallery').length === 0) { ... }
```

**Review by pxska — CHANGES_REQUESTED** [review]

File: sources/javascripts/concat/base.js [hunk 11]
```diff
@@ -537,6 +537,38 @@
...
+      changeProductImagePos();
+    }, 25));
+
+    var changeProductImagePos = function () {
+      var productGallery = $('.js-product-gallery');
+      var productImageContentBox = $('.js-content-item-box');
+      var buyBtnContent = $('.js-buy-btn-content');
+
+      if ($('.js-buy-btn-content .edy-buy-button-container').length >= 1) {
+        if ($(window).width() <= 640) {
```
Comment (pxska) [inline sources/javascripts/concat/base.js:559]: The width value should be dependant on the breakpoint of the design. The most accurate one here would be the [`$medium-down` breakpoint](https://github.com/Voog/design-santiago/blob/db264c000f039936c7e52357c1b5d4df9904bad8/sources/stylesheets/variables/_breakpoints.scss#L9) (as this is where the text goes under the product image.

**Review by tanelj — CHANGES_REQUESTED** [review]
Nice! Only some minor improvements.

File: sources/stylesheets/content-area/_text.scss (hunk 4 above)
Comment (robertpih) [reply sources/stylesheets/content-area/_text.scss:335]: Irrelevant.

File: sources/javascripts/concat/base.js (hunk 11 above)
Comment (tanelj) [reply sources/javascripts/concat/base.js]: One note. Since in JS we haven't breakpoint constant we must use number `640` but `<`

```suggestion
        if ($(window).width() < 640) {
```

File: sources/stylesheets/components/misc/_content-item.scss [hunk 12]
```diff
@@ -312,6 +362,24 @@
   box-sizing: border-box;
 }
 
+.product-price {
+  font-weight: 700;
+  font-size: 24px;
```
Comment (pxska) [inline sources/stylesheets/components/misc/_content-item.scss:367]: These values no longer seem necessary because they are now in the hands of the design editor.

File: sources/components/custom-styles/template-cs-style-rules.scss [hunk 13]
```diff
@@ -402,6 +403,15 @@
...
 
+// =============================================================================
+// Auto-rendered product layout
+// =============================================================================
+
+.product-price {
+  color: var(--main__color-primary);
+  font-family: var(--main__font-family);
+}
+
```
Comment (pxska) [inline sources/components/custom-styles/template-cs-style-rules.scss:414]: These values also seem irrelevant now.

**Review by tanelj — APPROVED** [review]
Greate job!

**Review by robertpih — APPROVED** [review]

**Review by tanelj — APPROVED** [review]

**Review by tanelj — APPROVED** [review]
